instrumentation/pubsub-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/pubsub/PubsubSingletons.java

```diff
@@ -107,22 +107,39 @@ public static void buildAndFinishSpan(Context context, PubsubMessage pubsubMessa
 
   public static Optional<Object> extractPubsubMessageAttributes(PubsubMessage pubsubMessage) {
     try {
-      Class cls = pubsubMessage.getClass();
-      Field attributes = cls.getDeclaredField(ATTRIBUTES_FIELD_NAME);
-      attributes.setAccessible(true);
-      Class attributesClass = attributes.get(pubsubMessage).getClass();
-      Field mapData = attributesClass.getDeclaredField(MAP_DATA_FIELD_NAME);
-      mapData.setAccessible(true);
-      Class mapDataObj = mapData.get(attributes.get(pubsubMessage)).getClass();
-
-      Field delegateField = mapDataObj.getDeclaredField(DELEGATE_DATA_FIELD_NAME);
-      delegateField.setAccessible(true);
-      return (Optional<Object>) delegateField.get(mapData.get(attributes.get(pubsubMessage)));
-
+      Field attributes = extractAttributeFromObject(pubsubMessage, ATTRIBUTES_FIELD_NAME);
+      if (attributes != null) {
+        Object attributesObject = attributes.get(pubsubMessage);
+        Field mapData = extractAttributeFromObject(attributesObject, MAP_DATA_FIELD_NAME);
+        if (mapData != null) {
+          Field delegate =
+              extractAttributeFromObject(mapData.get(attributesObject), DELEGATE_DATA_FIELD_NAME);
+          if (delegate != null) {
+            return (Optional<Object>) delegate.get(mapData.get(attributesObject));
+          }
+        }
+      }
     } catch (Exception e) {
-      System.out.println("Got Exception while instrumenting pubsubMessage: " + e);
-
+      return Optional.empty();
     }
     return Optional.empty();
   }
+
+  private static Field extractAttributeFromObject(Object object, String fieldName) {
```

Comment:
Why not return the property value itself?

Reply:
I thought its clearer to get the object from outside, no ?

Reply:
IMO this method is an abstraction for extracting the attribute, so we shouldn't be interested in reflection objects like Field, etc.

Reply:
ok. done.